diff --git a/cc/scheduler/scheduler_state_machine.cc b/cc/scheduler/scheduler_state_machine.cc
index 8cfa825188..7986bf196d 100644
--- a/cc/scheduler/scheduler_state_machine.cc
+++ b/cc/scheduler/scheduler_state_machine.cc
@@ -433,6 +433,10 @@ bool SchedulerStateMachine::ShouldDraw() const {
     return false;
   }

+  // Only draw if main thread committed this frame (no compositor-only frames).
+  if (!did_commit_during_frame_)
+    return false;
+
   // Browser compositor commit steals any resources submitted in draw. Therefore
   // drawing while a commit is pending is wasteful.
   if (settings_.commit_to_active_tree && CommitPending())
@@ -1389,6 +1393,7 @@ void SchedulerStateMachine::OnBeginImplFrame(const viz::BeginFrameArgs& args) {
   did_invalidate_layer_tree_frame_sink_ = false;
   did_perform_impl_side_invalidation_ = false;
   waiting_for_scroll_event_ = false;
+  needs_begin_main_frame_ = true;
 }

 void SchedulerStateMachine::OnBeginImplFrameDeadline() {
diff --git a/third_party/blink/renderer/platform/widget/compositing/layer_tree_settings.cc b/third_party/blink/renderer/platform/widget/compositing/layer_tree_settings.cc
index d41c0ab940..0b565e642c 100644
--- a/third_party/blink/renderer/platform/widget/compositing/layer_tree_settings.cc
+++ b/third_party/blink/renderer/platform/widget/compositing/layer_tree_settings.cc
@@ -569,8 +569,7 @@ cc::LayerTreeSettings GenerateLayerTreeSettings(
   settings.disallow_non_exact_resource_reuse = true;
 #endif

-  settings.wait_for_all_pipeline_stages_before_draw =
-      cmd.HasSwitch(::switches::kRunAllCompositorStagesBeforeDraw);
+  settings.wait_for_all_pipeline_stages_before_draw = true;

   settings.enable_image_animation_resync =
       !cmd.HasSwitch(switches::kDisableImageAnimationResync);
